Mathlib/Tactic/DeriveFintype.lean

```diff
@@ -0,0 +1,315 @@
+open Meta Command
+
+/-- Returns the portion of the proxy type for the constructor with arguments `xs`,
+the list of names used in the pattern, and the pattern for matching an element of this type. -/
+def mkCtorType (xs : List Expr) : TermElabM (Expr × List Name × TSyntax `term) :=
+  match xs with
+  | [] => return (mkConst ``Unit, [], ← `(term| ()))
+  | [x] => do
+    let a ← mkFreshUserName `a
+    return (← inferType x, [a], mkIdent a)
+  | x :: xs => do
+    let (ty, names, patt) ← mkCtorType xs
+    let ty ← mkAppM ``PSigma #[← mkLambdaFVars #[x] ty]
+    let a ← mkFreshUserName `a
+    return (ty, a :: names, ← `(term| ⟨$(mkIdent a), $patt⟩))
+
+/-- Navigates into the sum type that we create in `mkCType` for the given constructor index. -/
+def wrapSumAccess (cidx nctors : Nat) (spatt : TSyntax `term) : TermElabM (TSyntax `term) :=
+  match cidx with
+  | 0 =>
+    if nctors = 1 then
+      return spatt
+    else
+      `(term| Sum.inl $spatt)
+  | cidx' + 1 => do
+    let spatt ← wrapSumAccess cidx' (nctors - 1) spatt
+    `(term| Sum.inr $spatt)
+
+/-- Create a `Sum` of types, mildly optimized to not have a trailing `Empty`.
+Returns also the necessary tactic for the `left_inv` proof in the `Equiv` we construct. -/
+def mkCType (ctypes : List Expr) : TermElabM (Expr × TSyntax `tactic) :=
+  match ctypes with
+  | [] => return (mkConst ``Empty, ← `(tactic| cases x))
+  | [x] => return (x, ← `(tactic| rfl))
+  | x :: xs => do
+    let (ty, pf) ← mkCType xs
+    let pf ← `(tactic| cases x with | inl _ => rfl | inr x => $pf:tactic)
+    return (← mkAppM ``Sum #[x, ty], pf)
+
+structure EquivData where
+  /-- Name of the declaration for a type that is `Equiv` to the given type. -/
+  proxyName : Name
+  /-- A term that elaborates to something of type `proxyType ≃ type`.
+  Needs to be elaborated with this expected type. -/
+  equiv : TSyntax `term
+
+/--
+Generates a proxy type for the inductive type (defining it at `fintypeProxy` in the
+inductive type's namespace) and returns an equivalence to it.
+-/
+def mkProxyEquiv (indVal : InductiveVal) : TermElabM EquivData := do
+  if indVal.isRec then
+    throwError
+      "deriving Fintype: recursive inductive types are not supported (and are usually infinite)"
+  if 0 < indVal.numIndices then
+    throwError
+      "deriving Fintype: inductive indices are not supported"
+  let levels := indVal.levelParams.map Level.param
+  let proxyName := indVal.name.mkStr "fintypeProxy"
+  forallBoundedTelescope indVal.type indVal.numParams fun params _sort => do
+    let mut cdata := #[]
+    for ctorName in indVal.ctors do
+      let ctorInfo ← getConstInfoCtor ctorName
+      let ctorType ← inferType <| mkAppN (mkConst ctorName levels) params
+      cdata := cdata.push <| ←
+        forallBoundedTelescope ctorType ctorInfo.numFields fun xs _itype => do
+          let mut (ty, names, patt) ← mkCtorType xs.toList
+          if ← Meta.isProp ty then
+            -- We're using `Sum`, so insert a `PLift`.
```

Comment:
Any reason not to use `PSum`?

Reply:
It was just that there were no `Fintype` instances I could find for it.

For performance reasons, I've altered `mkCtorType` to always return a `Type _`. I noticed the `PSigma` instances all were defined by way of the `Sigma` instances via `Fintype.ofEquiv`, so to save some useless constructor wrangling it creates `Sigma`s directly, inserting `PLift` where needed.

Reply:
Ah, I think the answer is that `Fintype` is only allowed to apply to things in `Type _` not `Sort _`!